packages/piro/src/Piro_PerformAnalysis.cpp

```diff
@@ -748,6 +758,340 @@ if(useCustomDotProduct) {
+      ROL::Ptr<ROL::BoundConstraint<double> > bnd = ROL::makePtr<ROL::BoundConstraint_SimOpt<double> >(ROL::makePtrFromRef(u_bnd),boundConstraint);
+      ROL::Problem<double> prob(ROL::makePtrFromRef(obj), ROL::makePtrFromRef(sopt_vec));
+      prob.addBoundConstraint(bnd);
+      prob.addConstraint("Constraint", ROL::makePtrFromRef(constr),r_ptr);
+      bool lumpConstraints(false), printToStream(true);
+      prob.finalize(lumpConstraints, printToStream, *rolOutput);
+      ROL::Solver<double> optSolver(ROL::makePtrFromRef(prob), rolParams.sublist("ROL Options"));
+      optSolver.solve(*out);
+      return_status = optSolver.getAlgorithmState()->statusFlag;
+    } else {
+      *out << "Piro::PerformTROLAnalysis: Solving Full Space Unconstrained Optimization Problem" << std::endl;
+      ROL::Problem<double> prob(ROL::makePtrFromRef(obj), ROL::makePtrFromRef(sopt_vec));//, ROL::makePtrFromRef(constr), r_ptr);
+      prob.addConstraint("Constraint", ROL::makePtrFromRef(constr),r_ptr);
+      bool lumpConstraints(false), printToStream(true);
+      prob.finalize(lumpConstraints, printToStream, *rolOutput);
+      ROL::Solver<double> optSolver(ROL::makePtrFromRef(prob), rolParams.sublist("ROL Options"));
+      optSolver.solve(*out);
+      return_status = optSolver.getAlgorithmState()->statusFlag;
+    }
+    */
+  }
+  else {
+    ROL::ReducedDynamicObjective<double> reduced_obj(obj_ptr,constr_ptr,rol_x_ptr,rol_p_ptr,rol_lambda_ptr, *timeStamps, piroParams);
+
+    ROL::Ptr<ROL::ReducedDynamicObjective<double> > reduced_obj_ptr = ROL::makePtrFromRef(reduced_obj);
+    ROL::ReducedDynamicStationaryControlsObjective<double> reduced_stationarycontrols_obj(reduced_obj_ptr, rol_p_ptr, nt);
+
+    *out << "Piro::PerformTROLAnalysis: Solving Reduced Space Unconstrained Optimization Problem" << std::endl;
+    auto algo = ROL::TypeU::AlgorithmFactory<double>(rolParams.sublist("ROL Options"));
+    algo->run(rol_p_primal, reduced_stationarycontrols_obj, *rolOutput);
+    return_status = algo->getState()->statusFlag;
+    if (return_status == ROL::EExitStatus::EXITSTATUS_STEPTOL) return_status = 0;
+/*
+    if(boundConstrained) {
+      *out << "Piro::PerformTROLAnalysis: Solving Reduced Space Bound Constrained Optimization Problem" << std::endl;
+      auto algo = ROL::TypeB::AlgorithmFactory<double>(rolParams.sublist("ROL Options"));
+      algo->run(rol_p_primal, reduced_obj, *boundConstraint, *rolOutput); 
+      return_status = algo->getState()->statusFlag;
+    }  else {
+      *out << "Piro::PerformTROLAnalysis: Solving Reduced Space Unconstrained Optimization Problem" << std::endl;
+      auto algo = ROL::TypeU::AlgorithmFactory<double>(rolParams.sublist("ROL Options"));
+      algo->run(rol_p_primal, reduced_obj, *rolOutput);
+      return_status = algo->getState()->statusFlag;
+    }
+*/
+    //! check correctness of Gradient prvided by Model Evaluator
+    if(rolParams.get<bool>("Check Derivatives", false)) {
```

Comment:
It seems we are only checking the derivatives of the reduced objective. Is it possible to check also the partial derivatives?